codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -0,0 +1,280 @@
+            throw new IllegalArgumentException("Last stream id must not be set on GOAWAY frame");
+        }
+
+        int lastStreamId = http2Handler.connection().remote().lastStreamCreated()
+                           + frame.extraStreamIds() * 2;
+        http2Handler.goAway(
+                http2HandlerCtx, lastStreamId, frame.errorCode(), frame.content().retain(), promise);
+    }
+
+    private void writeStreamFrame(Http2StreamFrame frame, ChannelPromise promise) {
+        int streamId = frame.streamId();
+        if (frame instanceof Http2DataFrame) {
+            Http2DataFrame dataFrame = (Http2DataFrame) frame;
+            http2Handler.encoder().writeData(http2HandlerCtx, streamId, dataFrame.content().retain(),
+                                             dataFrame.padding(), dataFrame.isEndStream(), promise);
+        } else if (frame instanceof Http2HeadersFrame) {
+            Http2HeadersFrame headerFrame = (Http2HeadersFrame) frame;
+            http2Handler.encoder().writeHeaders(
+                    http2HandlerCtx, streamId, headerFrame.headers(), headerFrame.padding(), headerFrame.isEndStream(),
+                    promise);
+        } else if (frame instanceof Http2ResetFrame) {
+            Http2ResetFrame rstFrame = (Http2ResetFrame) frame;
+            http2Handler.resetStream(http2HandlerCtx, streamId, rstFrame.errorCode(), promise);
+        } else {
+            throw new UnsupportedMessageTypeException(frame);
+        }
+    }
+
+    private final class ConnectionListener extends Http2ConnectionAdapter {
+        @Override
+        public void onStreamActive(Http2Stream stream) {
+            if (ctx == null) {
+                // UPGRADE stream is active before handlerAdded().
+                return;
+            }
+            ctx.fireUserEventTriggered(new Http2StreamActiveEvent(stream.id()));
+        }
+
+        @Override
+        public void onStreamClosed(Http2Stream stream) {
+            ctx.fireUserEventTriggered(new Http2StreamClosedEvent(stream.id()));
+        }
+
+        @Override
+        public void onGoAwayReceived(final int lastStreamId, long errorCode, ByteBuf debugData) {
+            final Http2GoAwayFrame goAway = new DefaultHttp2GoAwayFrame(lastStreamId, errorCode, debugData);
+            ctx.fireChannelRead(goAway);
+        }
+    }
+
+    private final class InternalHttp2ConnectionHandler extends Http2ConnectionHandler {
+        InternalHttp2ConnectionHandler(Http2ConnectionDecoder decoder, Http2ConnectionEncoder encoder,
+                                       Http2Settings initialSettings) {
+            super(decoder, encoder, initialSettings);
+        }
+
+        @Override
+        protected void onStreamError(ChannelHandlerContext ctx, Throwable cause,
+                                     Http2Exception.StreamException http2Ex) {
+            try {
+                Http2Stream stream = http2Handler.connection().stream(http2Ex.streamId());
```

Comment:
instead of using `http2Handler` should we just use `this`? (`http2Handler.connection().stream(http2Ex.streamId())` -> `connection().stream(http2Ex.streamId())`), and make `InternalHttp2ConnectionHandler` a static class? This would simplify the cyclic dependency between this class and the outer class.